src/attributes/diagnostics.md

```diff
@@ -83,6 +85,56 @@ pub mod m3 {
 > [command-line][rustc-lint-cli], and also supports [setting
 > caps][rustc-lint-caps] on the lints that are reported.
 
+All lint attributes support an additional `reason` parameter, to give context why
+a certain attribute was added. This reason will be displayed as part of the lint
+message, if the lint is emitted at the defined level.
+
+```rust
+use std::path::PathBuf;
+
+pub fn get_path() -> PathBuf {
+    #[allow(unused_mut, reason = "this is only modified on some platforms")]
+    let mut file_name = PathBuf::from("git");
+
+    #[cfg(target_os = "windows")]
+    file_name.set_extension("exe");
+
+    file_name
+}
+```
+
+### Lint expectations
+
+With the `#[expect]` attributes lints can be expected in a certain scope. If
+this expectation is not fulfilled a new warning is emitted to the user. The
+lint levels can be overridden with other lint attributes as usual.
+
+```rust
+#[warn(missing_docs)]
+pub mod m2{
+    #[expect(missing_docs)]
+    pub mod nested {
+        // This missing documentation fulfills the expectation above
+        pub fn undocumented_one() -> i32 { 1 }
+
+        // Missing documentation signals a warning here, despite the expectation
+        // above. This emission would not fulfill the expectation
+        #[warn(missing_docs)]
+        pub fn undocumented_two() -> i32 { 2 }
+    }
+
+    #[expect(missing_docs)]
+    /// This comment explains something cool about the function. The
+    /// expectation will not be fulfilled and in turn issue a warning.
+    pub fn undocumented_too() -> i32 { 3 }
+}
+```
+
+> Note: Lint expectations have been proposed in [RFC 2383]. It was not defined
+> how expectations of the expectation lint should be handled. The rustc
+> implementation currently doesn't allow the expextation of the
+> `unfulfilled_lint_expectation` lint. This can change in the future.
```

Comment:
I'm not sure the link to the RFC is particularly important here.  There are also a few typos here (`unfulfilled_lint_expectation` (no s) and `expextation`). It is also not clear to me what it means by "currently doesn't allow", as it seems to allow it just fine:

```rust
#![feature(lint_reasons)]
#![expect(unfulfilled_lint_expectations)]
```

Produces:

```
warning: this lint expectation is unfulfilled
 --> src/lib.rs:2:11
  |
2 | #![expect(unfulfilled_lint_expectations)]
  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default
  = note: the `unfulfilled_lint_expectations` lint can't be expected and will always produce this message
```

It does seem like it is *always* a warning, even if `unfulfilled_lint_expectations` is triggered (which seems strange).
Is there some more context here? 

Perhaps it could say something like:

> **Note**: The behavior of `#[expect(unfulfilled_lint_expectations)]` is currently defined to always generate the `unfulfilled_lint_expectations` lint. This may change in the future.

Perhaps this should be clarified before it is stabilized?

Reply:
> Note: The behavior of `#[expect(unfulfilled_lint_expectations)]` is currently defined to always generate the `unfulfilled_lint_expectations` lint. This may change in the future.

Yes this is what I intended to say. The context sadly cannot be condensed to one link. This behavior was implemented in https://github.com/rust-lang/rust/pull/94670. This has been discussed in the RFC PR (https://github.com/rust-lang/rfcs/pull/2383#issuecomment-396794262) with the suggestion to allow the `unfulfilled_lint_expectations` to be caught by outer expectations.

This can also be implemented in rustc. However, I decided against this for two reasons:
1. I didn't see a benefit. In the RFC Discussion, it was also said that there seems to be no use case for `#[expect(unfulfilled_lint_expectations)]`
2. Supporting `#[expect(unfulfilled_lint_expectations)]` would add a bunch of complexity. Currently, the lint is emitted at the item the attribute is attached to (For reasons regarding, incremental compilation and hash stability). To support such an expectation, we would need to either store the lint level and then order the expectations by which *could* fulfill which one or depend on the processing order which has caused some issues in the past. So, either solution would add complexity to the implementation without having an use case. We can also not simply emit the lint at the parent item due to the way lint levels are processed.

---

btw how did you add the `(i) Note` marker? I can sadly not see the MD text